### PR TITLE
CEFR level estimation & single word translation feature

### PR DESCRIPTION
**Pull request by Akhunzianov**

### CODE REVIEW

File: backend/src/main/java/com/lingvoFriend/backend/Services/AuthService/models/UserModel.java [hunk 1]
```diff
@@ -37,6 +37,9 @@ public class UserModel {
     private List<String> interests;
     private QuestionState questionState = QuestionState.NOT_STARTED;
 
+    private Integer levelEvaluationQuestionsAsked = 0;
+    private String cefrLevel;
```
Comment (sharkov63) [inline backend/src/main/java/com/lingvoFriend/backend/Services/AuthService/models/UserModel.java:41]: nit Maybe we could use `enum` here?

File: frontend/src/components/pages/chat/Chat.jsx [hunk 2]
```diff
@@ -85,6 +100,92 @@ const Chat = ({ username }) => {
...
+        if (translations[word]) {
+            return;
+        }
+
+        setLoadingWords((prev) => ({ ...prev, [word]: true }));
+
+        try {
+            const response = await axios.post("https://api-free.deepl.com/v2/translate", null, {
+                params: {
+                    auth_key: "805412aa-0cfc-4096-b255-74aaf6f8fbae:fx",
```
[the remark this answers is not in the document]
Comment (Akhunzianov) [reply frontend/src/components/pages/chat/Chat.jsx:113]: там в целом бесплатный API, но пофикшу энивэй

File: backend/src/main/java/com/lingvoFriend/backend/Services/ChatService/ChatService.java [hunk 3]
```diff
@@ -36,4 +37,80 @@ public List<Message> getMessagesByUsername(String username) {
...
+    public boolean isEvaluatingFinished(String username) {
+        Optional<UserModel> user = userRepository.findByUsername(username);
+        if (user.isPresent()) {
+            Integer status = user.get().getLevelEvaluationQuestionsAsked();
+            return status >= levelEvalQuestionsNumber;
+        }
+        throw new BadCredentialsException("User not found");
+    }
+
+    public boolean isEvaluatinCEFR(String username) {
```
Comment (sharkov63) [inline backend/src/main/java/com/lingvoFriend/backend/Services/ChatService/ChatService.java:59]: typo: `isEvaluatingCEFR`

File: backend/src/main/java/com/lingvoFriend/backend/Controllers/MainController.java [hunk 4]
```diff
@@ -46,20 +46,88 @@ public Mono<ResponseEntity<String>> sendRequestToLLM(@RequestBody LlmRequestDto
...
+        if (!(Objects.equals(userMessage.getRole(), "user")
+                || (Objects.equals(userMessage.getRole(), "system")))) {
             return Mono.just(
                     ResponseEntity.status(HttpStatus.BAD_REQUEST).body("Bad message role"));
         }
 
+        List<Message> updatedMessages;
+
+        chatService.addMessageToUser(username, userMessage);
+        if (!chatService.isEvaluating(username) && !chatService.isEvaluatingFinished(username)) {
```
Comment (sharkov63) [inline backend/src/main/java/com/lingvoFriend/backend/Controllers/MainController.java:61]: It seems that the chat controller contains logic about user evaluation process and prompt. We should extract this logic to a new service, say `EvaluationService`

File: backend/src/main/java/com/lingvoFriend/backend/Services/ChatService/ChatService.java [hunk 5]
```diff
@@ -36,4 +37,80 @@ public List<Message> getMessagesByUsername(String username) {
...
+        Optional<UserModel> user = userRepository.findByUsername(username);
+        if (user.isPresent()) {
+            Integer status = user.get().getLevelEvaluationQuestionsAsked();
+            return status == levelEvalQuestionsNumber;
+        }
+        throw new BadCredentialsException("User not found");
+    }
+
+    public void questionsAskedPlusOne(String username) {
```
Comment (sharkov63) [inline backend/src/main/java/com/lingvoFriend/backend/Services/ChatService/ChatService.java:68]: nit `incrementEvaluationQuestionsAsked`?

File: backend/src/main/java/com/lingvoFriend/backend/Services/ChatService/ChatService.java [hunk 6]
```diff
@@ -36,4 +37,80 @@ public List<Message> getMessagesByUsername(String username) {
...
+    }
+
+    public boolean isEvaluatinCEFR(String username) {
+        Optional<UserModel> user = userRepository.findByUsername(username);
+        if (user.isPresent()) {
+            Integer status = user.get().getLevelEvaluationQuestionsAsked();
+            return status == levelEvalQuestionsNumber;
+        }
+        throw new BadCredentialsException("User not found");
+    }
```
Comment (virogg) [inline backend/src/main/java/com/lingvoFriend/backend/Services/ChatService/ChatService.java:66]: В чем глобальная разница методов `isEvaluatingFinished` и `isEvaluatinCEFR`?

**Comment by sharkov63** [comment]
Converted to #84 and #86

**Pull request closed without merging**